src/modules/tpm/src/lib/Tpm.cpp

```diff
@@ -49,37 +51,82 @@ int Tpm::Get(const char* objectName, MMI_JSON_STRING* payload, int* payloadSizeB
     int status = MMI_OK;
     std::string data;
 
-    if (std::strcmp(objectName, TPM_STATUS) == 0)
-    {
-        GetStatus(data);
-    }
-    else if (std::strcmp(objectName, TPM_VERSION) == 0)
-    {
-        GetVersion(data);
-    }
-    else if (std::strcmp(objectName, TPM_MANUFACTURER) == 0)
+    if ((std::strcmp(objectName, TPM_STATUS) == 0) || (std::strcmp(objectName, TPM_VERSION) == 0) || (std::strcmp(objectName, TPM_MANUFACTURER) == 0))
     {
-        GetManufacturer(data);
+        if (std::strcmp(objectName, TPM_STATUS) == 0)
+        {
+            GetStatus(data);
+        }
+        else if ((std::strcmp(objectName, TPM_VERSION) == 0) && (this->m_hasCapabilitiesFile))
+        {
+            GetVersionFromCapabilitiesFile(data);
+        }
+        else if ((std::strcmp(objectName, TPM_MANUFACTURER) == 0) && (this->m_hasCapabilitiesFile))
+        {
+            GetManufacturerFromCapabilitiesFile(data);
+        }
     }
     else
     {
+        if (IsFullLoggingEnabled())
+        {
+            OsConfigLogError(TpmLog::Get(), "Invalid argument, objectName %s not found", objectName);
+        }
         status = EINVAL;
     }
 
+    if ((2 >= data.length()) && (status == MMI_OK) && ((0 == std::strcmp(objectName, TPM_VERSION)) || (0 == std::strcmp(objectName, TPM_MANUFACTURER))))
+    {
+        this->m_hasCapabilitiesFile = false;
+        std::string tpmProperty;
+        status = Tpm2Utils::GetTpmPropertyFromDeviceFile(objectName, tpmProperty);
+        if ((status == MMI_OK) && (!tpmProperty.empty()))
+        {
+            if (0 == std::strcmp(objectName, TPM_VERSION))
+            {
+                std::regex re(g_tpmVersionFromDeviceFile);
+                std::smatch match;
+                if (std::regex_search(tpmProperty, match, re))
+                {
+                    data = char('"') + tpmProperty + char('"');
+                }
+            }
+            else if (0 == std::strcmp(objectName, TPM_MANUFACTURER))
+            {
+                std::regex re(g_tpmManufacturerFromDeviceFile);
+                std::smatch match;
+                if (std::regex_search(tpmProperty, match, re))
+                {
+                    data = char('"') + tpmProperty + char('"');
+                }
+            }
+        }
+        else
+        {
+            this->m_hasCapabilitiesFile = true;
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogError(TpmLog::Get(), "Tpm property for object %s not found", objectName);
+            }
+        }
+    }
+
     data.erase(std::find(data.begin(), data.end(), '\0'), data.end());
 
-    if (!((this->m_maxPayloadSizeBytes > 0) && ((int)data.length() > this->m_maxPayloadSizeBytes)) &&
-        (status == MMI_OK))
+    if ((((int)data.length() < this->m_maxPayloadSizeBytes) || (this->m_maxPayloadSizeBytes == 0)) && (status == MMI_OK))
```

Comment:
Should we use `C++` style cast here?

Reply:
Great point. In this case, this->m_maxPayloadSizeBytes actually had an incorrect underlying type declaration ('int', when it should be 'const unsigned int'). This is now corrected and there is no longer any need to cast. Thank you for catching this!